org.jacoco.agent.rt/src/org/jacoco/agent/rt/internal/PreMain.java

```diff
@@ -58,31 +61,46 @@ public static void premain(final String options, final Instrumentation inst)
 	private static IRuntime createRuntime(final Instrumentation inst)
 			throws Exception {
 
-		if (redefineJavaBaseModule(inst)) {
-			return new InjectedClassRuntime(Object.class, "$JaCoCo");
+		IRuntime injectedClassRuntime = createInjectedClassRuntime(inst);
+		if (injectedClassRuntime != null) {
+			return injectedClassRuntime;
 		}
 
 		return ModifiedSystemClassRuntime.createFor(inst,
 				"java/lang/UnknownError");
 	}
 
-	/**
-	 * Opens {@code java.base} module for {@link InjectedClassRuntime} when
-	 * executed on Java 9 JREs or higher.
-	 *
-	 * @return <code>true</code> when running on Java 9 or higher,
-	 *         <code>false</code> otherwise
-	 * @throws Exception
-	 *             if unable to open
-	 */
-	private static boolean redefineJavaBaseModule(
+	private static IRuntime createInjectedClassRuntime(
 			final Instrumentation instrumentation) throws Exception {
 		try {
 			Class.forName("java.lang.Module");
 		} catch (final ClassNotFoundException e) {
-			return false;
+			return null;
+		}
+		IRuntime runtime = (IRuntime) (new ClassLoader() {
+			@Override
+			public Class<?> loadClass(String name)
+					throws ClassNotFoundException {
+				if (!name.startsWith(InjectedClassRuntime.class.getName())) {
+					return super.loadClass(name);
+				}
+				final InputStream resourceAsStream = getResourceAsStream(
+						name.replace('.', '/') + ".class");
+				final byte[] bytes;
+				try {
+					bytes = InputStreams.readFully(resourceAsStream);
+				} catch (IOException e) {
+					throw new RuntimeException(e);
+				}
+				return defineClass(name, bytes, 0, bytes.length);
+			}
+		}.loadClass(InjectedClassRuntime.class.getName())
+				.getConstructor(Class.class, String.class)
+				.newInstance(Object.class, "$JaCoCo"));
+		final Object module = getModule(runtime.getClass());
+		if (module.equals(getModule(PreMain.class))) {
+			throw new IllegalStateException();
 		}
-
 		Instrumentation.class.getMethod("redefineModule", //
```

Comment:
Maybe for consistency and better readability this code can be extracted to a private method like `getModule()`